### DISCUSSION TITLE
Proposal: A generic type indicating a primitive arithmetic type.

### DISCUSSION THREAD

**Discussion by Monsoonexe** [dotnet/csharplang#5987] [Language Ideas]
When writing functions where the logic is the same but the data type is a different primitive arithemetic type, I often find myself copy-pasting functions just to change the underlying type. Below is a trivial example, but illustrates that this function must be repeated to change the return type and accumulator type.

For example: 
```
    public static int Sum(this IList<int> col)
    {
        int sum = 0; //return type 'int'
        for (int i = col.Count - 1; i >= 0; --i)
            sum += col[i];
        return sum;
    }

    public static long Sum(this IList<long> col)
    {
        long sum = 0; //return type 'long'
        for (int i = col.Count - 1; i >= 0; --i)
            sum += col[i];
        return sum;
    }

    public static ulong Sum(this IList<ulong> col)
    {
        ulong sum = 0; //return type 'ulong'
        for (int i = col.Count - 1; i >= 0; --i)
            sum += col[i];
        return sum;
    }

//etc... for short, ushort, float, double

public static void ImplementSum()
{
     int[] intArray = new int[] {1, 2, 3 };
     int sum = intArray.Sum(); //Sum<int>();
}
```

I propose a generic 'IArithmetic' interface for primitives that support the arithmetic operators like "+, -, *, /", such that the following is valid syntax:
```
public static TArithmetic Sum(this IList<TArithmetic> col)
where TArithmetic : IArithmetic
{
        TArithmetic sum = 0; //primitive type that can +, -, *, /
        for (int i = col.Count - 1; i >= 0; --i)
            sum += col[i];
        return sum;
}

public static void ImplementSum()
{
    int[] intArray = new int[] {1, 2, 3};
    int six = intArray .Sum(); //<int> is implicit because int implements 'IArithmetic'

    ulong[] ulongArray = new ulong[] {1, 2, 3};
    ulong uSix = ulongArray .Sum<ulong>();//<ulong> is implicit because uint implements 'IArithmetic'
}
```

Arithemetic types are any primitive that support the arithmetic operators like "+, -, *, /". 
Similarly to how some primitives implement the IComparable interface, they would also implement IArithmetic.
The goal of this feature is to reduce repetition when working with primitive datatypes.

An IArithmetic interface also opens the possibility for classes to also inherit from IArithmetic if they can implement all the same operators, in the same way classes can be IComparable and are used with the ">, <, ==" (comparison) operators. 

I imagine this would be pretty hard to implement due to the sizes of the underlying data types. There could be a "when in doubt, promote" policy to help the compiler make the decisions about which type to use.

**Comment by ufcpp** [comment]
Generic math?

see https://devblogs.microsoft.com/dotnet/preview-features-in-net-6-generic-math/

**Comment by 333fred** [comment]
Well, do I have the blog post (and preview feature) for you: https://devblogs.microsoft.com/dotnet/preview-features-in-net-6-generic-math/
